---
title: "FINEMAP-MISS"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{FINEMAP-MISS}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Data
Using data from 1000 genomes <ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr6.phase3_shapeit2_mvncall_integrated_v5b.20130502.genotypes.vcf.gz>, a example GWAS was generated for two cohorts. Variant 99 has a true effect on the simulated phenotype, explaining 1\% of the phenotypic variance. 

```{r setup}
library(FINEMAPMISS)
knitr::opts_chunk$set(fig.width=7, fig.height=5, out.width="90%")
```

```{r read data}
data("toydata_FINEMAPMISS")
betas <- toydata_FINEMAPMISS$betas
ses <- toydata_FINEMAPMISS$ses
MAF <- toydata_FINEMAPMISS$MAF
LD <- toydata_FINEMAPMISS$LD

n <- toydata_FINEMAPMISS$study_sample_sizes
p <- dim(LD)[1]

par(mfrow = c(1,2))
minus_log10p <- -log10(2*(1 - pnorm(abs(betas/ses))))
plot(minus_log10p[,1], ylab = "-log10P", main = "Study 1")
points(toydata_FINEMAPMISS$causal_snp, minus_log10p[toydata_FINEMAPMISS$causal_snp,1],
       pch = 21, bg = "green")
legend("topleft", pch = 21, legend = c("Causal variant"), pt.bg = "green")
plot(minus_log10p[,2], ylab = "-log10P", main = "Study 2")
points(toydata_FINEMAPMISS$causal_snp, minus_log10p[toydata_FINEMAPMISS$causal_snp,2],
       pch = 21, bg = "green")

```

Let's simulate missing data by removing values for 20\% of variants from one study.


```{r simulate missing data}
set.seed(1)

#Simulating missingness in 20% of variants (including the true causal variant)
missing_data <- unique(sort(c(toydata_FINEMAPMISS$causal_snp,
                              cbind(sample(1:p, round(p*0.2))))))

#Which dataset are the variants missing from?
which_dataset_missing <- sample(1:2, length(missing_data), replace = TRUE)

#Setting missing data to 0 or Inf.
betas[missing_data[which_dataset_missing == 1],1] <- 0
betas[missing_data[which_dataset_missing == 2],2] <- 0
ses[missing_data[which_dataset_missing == 1],1] <- Inf
ses[missing_data[which_dataset_missing == 2],2] <- Inf

#Setting variants sample sizes
variant_sample_sizes <- rep(sum(n), p)
variant_sample_sizes[missing_data] <- n[1]
```

Let's meta-analyze the summary statistics and visualize the p-values of the combined effects.
```{r meta_p}
meta_analysis <- IVW(betas = betas, ses = ses)
meta_minus_log10p <- -log10(2*(1 - pnorm(abs(meta_analysis$`beta meta`/meta_analysis$`se meta`))))
plot(meta_minus_log10p, ylab = "-log10P", main = "Meta-analysis")
points(toydata_FINEMAPMISS$causal_snp, meta_minus_log10p[toydata_FINEMAPMISS$causal_snp],
       pch = 21, bg = "green")
legend("topleft", pch = 21, legend = c("Causal variant"), pt.bg = "green")
```

# Fine-mapping
If we try to fine-map this partially missing data with FINEMAP, we cannot detect the true causal variant.
```{r FINEMAP}

FM_output <- original_FINEMAP(ses = ses, betas = betas, R = LD, n_studies = 2,
                              freqs = MAF, variant_sample_sizes = variant_sample_sizes)
plot(FM_output$summary_table$prob, ylab = "PIP", main = "FINEMAP PIPs")
points(toydata_FINEMAPMISS$causal_snp, FM_output$summary_table$prob[toydata_FINEMAPMISS$causal_snp],
       pch = 21, bg = "green")
points(FM_output$credible_sets[,1], FM_output$summary_table$prob[FM_output$cred[,1]],
       col = "red")
legend("topleft", pch = c(21,1), legend = c("Causal variant", "Credible set"), pt.bg = c("green", NA),
       col = c("black", "red"))
```

## FINEMAP-MISS
With FINEMAP-MISS, the true causal variant is included in the credible set.
```{r FINEMAP-MISS}

FMM_output <- run_FINEMAPMISS(ses = ses, betas = betas, R = LD, 
                          n_studies = 2, freqs = MAF, 
                          variant_sample_sizes = variant_sample_sizes,
                          scaled_data = T)
plot(FMM_output$summary_table$prob, ylab = "PIP", main = "FINEMAP-MISS PIPs")
points(toydata_FINEMAPMISS$causal_snp, FMM_output$summary_table$prob[toydata_FINEMAPMISS$causal_snp],
       pch = 21, bg = "green")
points(FMM_output$credible_sets[,1], FMM_output$summary_table$prob[FMM_output$credible_sets[,1]],
       col = "red")
legend("topleft", pch = c(21,1), legend = c("Causal variant", "Credible set"), pt.bg = c("green", NA),
       col = c("black", "red"))
```

## Summary statistics imputation
We can alternatively impute missing data and then run FINEMAP.
````{r FINEMAP imputed}
#Setting variants sample sizes
variant_sample_sizes <- rep(sum(n), p)
variant_sample_sizes[missing_data] <- n[1]

#Index of observed variants from each study
obs1 <- setdiff(1:p, missing_data[which_dataset_missing == 1])
obs2 <- setdiff(1:p, missing_data[which_dataset_missing == 2])

#'#Index of unobserved variants from each study
unobs1 <- sort(missing_data[which_dataset_missing == 1])
unobs2 <- sort(missing_data[which_dataset_missing == 2])

#z-scores from each study
z_obs1 <- (betas/ses)[obs1, 1]
z_obs2 <- (betas/ses)[obs2, 2]

imputation1 <- impute_summary_stats(R = LD, z = z_obs1,
                                    observed = obs1,
                                    unobserved = unobs1, n = n[1],
                                    return_z = FALSE, scale = TRUE)


imputation2 <- impute_summary_stats(R = LD, z = z_obs2,
                                   observed = obs2,
                                   unobserved = unobs2, n = n[2],
                                   return_z = FALSE, scale = TRUE)

#Copying over summary statistics
betas_imputed <- betas
ses_imputed <- ses

#Using imputed values to replace missing observations
betas_imputed[unobs1,1] <- imputation1[,1]
ses_imputed[unobs1,1] <- imputation1[,3]

betas_imputed[unobs2,2] <- imputation2[,1]
ses_imputed[unobs2,2] <- imputation2[,3]

#Assuming all variants fully observed after imputation
variant_sample_sizes <- rep(sum(n),p)

#Running FINEMAP
FM_imputed_output <- original_FINEMAP(ses = ses_imputed,
                         betas = betas_imputed,
                         R = LD,
                         n_studies = 2,
                         variant_sample_sizes = variant_sample_sizes,
                         freqs = toydata_FINEMAPMISS$MAF, scaled_data = T)

plot(FM_imputed_output$summary_table$prob, ylab = "PIP", main = "FINEMAP PIPs")
points(toydata_FINEMAPMISS$causal_snp, FM_imputed_output$summary_table$prob[toydata_FINEMAPMISS$causal_snp],
       pch = 21, bg = "green")
points(FM_imputed_output$cred[,1], FM_imputed_output$summary_table$prob[FM_imputed_output$credible_sets[,1]],
       col = "red")
legend("topleft", pch = c(21,1), legend = c("Causal variant", "Credible set"), pt.bg = c("green", NA),
       col = c("black", "red"))
````




